x/fswap/keeper/keeper.go

```diff
@@ -6,37 +6,217 @@ import (
 	"github.com/tendermint/tendermint/libs/log"
 
 	"github.com/Finschia/finschia-sdk/codec"
+	"github.com/Finschia/finschia-sdk/store/prefix"
 	storetypes "github.com/Finschia/finschia-sdk/store/types"
 	sdk "github.com/Finschia/finschia-sdk/types"
+	sdkerrors "github.com/Finschia/finschia-sdk/types/errors"
 	"github.com/Finschia/finschia-sdk/x/fswap/types"
 )
 
-type (
-	Keeper struct {
-		cdc           codec.BinaryCodec
-		storeKey      storetypes.StoreKey
-		accountKeeper types.AccountKeeper
-		bankKeeper    types.BankKeeper
-		config        types.Config
-	}
-)
+type Keeper struct {
+	cdc      codec.BinaryCodec
+	storeKey storetypes.StoreKey
+
+	config types.Config
+
+	AccountKeeper
+	BankKeeper
+}
 
-func NewKeeper(
-	cdc codec.BinaryCodec,
-	storeKey storetypes.StoreKey,
-	ak types.AccountKeeper,
-	bk types.BankKeeper,
-	config types.Config,
-) Keeper {
+func NewKeeper(cdc codec.BinaryCodec, storeKey storetypes.StoreKey, config types.Config, ak AccountKeeper, bk BankKeeper) Keeper {
 	return Keeper{
-		cdc:           cdc,
-		storeKey:      storeKey,
-		accountKeeper: ak,
-		bankKeeper:    bk,
-		config:        config,
+		cdc,
+		storeKey,
+		config,
+		ak,
+		bk,
 	}
 }
 
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
+
+func (k Keeper) Swap(ctx sdk.Context, addr sdk.AccAddress, fromCoinAmount sdk.Coin, toDenom string) error {
+	swap, err := k.getSwap(ctx, fromCoinAmount.Denom, toDenom)
+	if err != nil {
+		return err
+	}
+
+	newAmount := fromCoinAmount.Amount.Mul(swap.SwapMultiple)
+	newCoinAmount := sdk.NewCoin(toDenom, newAmount)
+	if err := k.checkSwapCap(ctx, swap, newCoinAmount); err != nil {
```

Comment:
Both of `checkSwapCap()` and `updateSwapped()` call `getSwapped()` within its logic. It would be nice to refactor relevant flow. e.g. get swapped -> update it -> validate the result -> set swapped.